akka-stream/src/main/scala/akka/stream/impl/io/IOSources.scala

```diff
@@ -4,46 +4,151 @@
+      attachment(Failure(ex))
+    }
   }
+}
 
-  override protected def newInstance(shape: SourceShape[ByteString]): SourceModule[ByteString, Future[IOResult]] =
-    new FileSource(f, chunkSize, attributes, shape)
+/**
+ * INTERNAL API
+ * Creates simple asynchronous Source backed by the given file.
+ */
+private[akka] final class FileSource(path: Path, chunkSize: Int)
+  extends GraphStageWithMaterializedValue[SourceShape[ByteString], Future[IOResult]] {
+  require(chunkSize > 0, "chunkSize must be greater than 0")
 
-  override def withAttributes(attr: Attributes): Module =
-    new FileSource(f, chunkSize, attr, amendShape(attr))
+  val out = Outlet[ByteString]("FileSource.out")
+
+  override protected def initialAttributes: Attributes = DefaultAttributes.fileSource
+
+  override val shape = SourceShape(out)
+
+  override def createLogicAndMaterializedValue(inheritedAttributes: Attributes): (GraphStageLogic, Future[IOResult]) = {
+    val ioResultPromise = Promise[IOResult]()
+
+    val logic = new TimerGraphStageLogic(shape) with OutHandler {
+      val maxReadAhead = inheritedAttributes.getAttribute(classOf[InputBuffer], InputBuffer(16, 16)).max
+
+      val bufferSize = if (maxReadAhead > 2) (chunkSize * maxReadAhead) / 2 else chunkSize
+      val buffer = ByteBuffer.allocate(bufferSize)
+
+      var channel: AsynchronousFileChannel = _
+      var position = 0L
+      var chunkCallback: Try[Int] ⇒ Unit = _
+      var readBytes = -1
+      var waitingForData = false
+
+      private def dataIsReady = readBytes != -1
+      private def setDataReady(bytes: Int) = readBytes = bytes
+
+      override def preStart(): Unit = {
+        try {
+          // this is a bit weird but required to keep existing semantics
+          require(Files.exists(path), s"Path '$path' does not exist")
+          require(Files.isRegularFile(path), s"Path '$path' is not a regular file")
+          require(Files.isReadable(path), s"Missing read permission for '$path'")
+        } catch {
+          case ex: IllegalArgumentException ⇒
+            ioResultPromise.trySuccess(IOResult(position, Failure(ex)))
+            throw ex
+        }
+        channel = AsynchronousFileChannel.open(path, StandardOpenOption.READ)
+
+        chunkCallback = getAsyncCallback[Try[Int]] {
+          case Success(readBytes) ⇒
+            waitingForData = false
+            setDataReady(readBytes)
+            if (isAvailable(out)) pushBuffer()
+
+          case Failure(ex) ⇒
+            failStage(ex)
+            ioResultPromise.trySuccess(IOResult(position, Failure(ex)))
+
+        }.invoke
+      }
+
+      private def pushBuffer(): Unit = {
+        if (readBytes > 0) {
+          buffer.flip()
+          val up = math.ceil(readBytes.toDouble / chunkSize).toInt
+          val list = for (_ ← 1 to up) yield {
+            val arr = new Array[Byte](if (buffer.remaining > chunkSize) chunkSize else buffer.remaining)
```

Comment:
can't we reuse a single Array instance and use `buffer.get` with offset and length params?